### DISCUSSION TITLE
How to draw to desktop while picom is running

### DISCUSSION THREAD

**Discussion by scrouthtv** [yshui/picom#643] [Q&A] [answered]
This is not a bug report, but an off-topic programming question. Feel free to close if it's too far off-topic.

I am currently trying to write a program to set the wallpaper. From what I can tell, this is basically done by drawing to X's root window. I have this program so far:

```c
#include <stdlib.h>
#include <stdio.h>
#include <xcb/xcb.h>

int main () {
	xcb_rectangle_t rectangles[] = {
		{ 200, 200, 400, 400 },
	};

	xcb_connection_t *c = xcb_connect(NULL, NULL);

	/* get the first screen */
	xcb_screen_t *screen = xcb_setup_roots_iterator(xcb_get_setup(c)).data;

	xcb_gcontext_t gc = xcb_generate_id(c);

	/* root window */
	xcb_drawable_t draw = screen->root;
	/* or any other window id: xcb_drawable_t draw = 0x4200004; */

	uint32_t mask = XCB_GC_FUNCTION | XCB_GC_FOREGROUND | XCB_GC_BACKGROUND |	XCB_GC_LINE_WIDTH| XCB_GC_LINE_STYLE | XCB_GC_GRAPHICS_EXPOSURES;
	uint32_t values[] = {
			XCB_GX_XOR,
			screen->white_pixel,
			screen->black_pixel,
			1,
			XCB_LINE_STYLE_ON_OFF_DASH,
			0
	};

	xcb_create_gc(c, gc, draw, mask, values);
	xcb_map_window(c, draw);
	while (true) {
		xcb_poly_rectangle(c, draw, gc, 3, rectangles);
		xcb_flush(c);
	}
	return 0;
}
```

This only works if picom is not running.

I tried setting the drawable directly to something else (see the comment below xcb_drawable_t draw = ...), which also didn't work.

Can someone give me a hint about how I'd go with drawing to the desktop while picom is running?
Thanks in advance.

**Comment by yshui** [comment, marked as answer]
Wallpaper is usually set by setting the background pixmap of the X root window.

picom doesn't render the content of the root window, so anything that was drawn on the root window will be lost when picom is running.

**Reply by scrouthtv** [reply]
How would I draw the wallpaper if picom is running if drawing to the root window does not work?

**Reply by yshui** [reply]
> Wallpaper is usually set by setting the background pixmap of the X root window.

**Reply by scrouthtv** [reply]
But
> anything that was drawn on the root window will be lost when picom is running

or is there a difference between the background pixmap and the window content itself?
How would I draw to the background pixmap?

**Reply by yshui** [reply]
https://tronche.com/gui/x/xlib/window/XSetWindowBackgroundPixmap.html https://linux.die.net/man/1/xsri

**Reply by scrouthtv** [reply]
Hmm, I tried drawing to the desktop using an [answer from StackOverflow](https://stackoverflow.com/a/15000011/7242251).

It also only works if picom is not running - if I start picom, the background image is gone.
Even re-running the program does not render the background.

**Reply by scrouthtv** [reply]
Any idea @yshui
